aptos/contract/aux/sources/stake.move

```diff
@@ -0,0 +1,1571 @@
+            if (pool.reward_remaining == 0) {
+                pool.end_time = now;
+            }
+        };
+        // the real balance of the reward coin should always be >= the virtual remaining balance
+        assert!(coin::value(&pool.reward) >= pool.reward_remaining, E_INTERNAL_ERROR);
+
+        if (time_amount_us > 0 && time_increase) {
+            // If the pool has expired, start the reward period from now
+            if (now >= pool.end_time) {
+                pool.start_time = now;
+                pool.end_time = now + time_amount_us;
+            } else {
+                pool.end_time = pool.end_time + time_amount_us;
+            }
+        // TODO: should we allow this?
+        } else if (time_amount_us > 0 && !time_increase) {
+            assert!(time_amount_us < pool.end_time, E_INVALID_DURATION);
+            pool.end_time = pool.end_time - time_amount_us;
+            assert!(pool.end_time > now, E_INVALID_DURATION);
+        };
+        // If any reward remains undistributed, the end time must be in the
+        // future so that the remaining reward can be distributed
+        if (pool.reward_remaining > 0) {
+            assert!(pool.end_time > now, E_INVALID_DURATION);
+        };
+        if (pool.end_time > now) {
+            assert!(pool.end_time - pool.start_time <= MAX_DURATION_US, E_INVALID_DURATION);
+            assert!(pool.end_time - pool.start_time >= MIN_DURATION_US, E_INVALID_DURATION);
+        };
+
+        event::emit_event<ModifyPoolEvent<S, R>>(
+            &mut pools.modify_pool_events,
+            ModifyPoolEvent {
+                pool_id: id,
+                authority: pool.authority,
+                start_time: pool.start_time,
+                end_time: pool.end_time,
+                reward_remaining: pool.reward_remaining,
+                total_amount_staked: coin::value(&pool.stake),
+                acc_reward_per_share: pool.acc_reward_per_share
+            }
+        );
+    }
+
+    fun update_pool<S, R>(pool: &mut Pool<S, R>, now: u64) {
+        if (now <= pool.last_update_time || pool.last_update_time >= pool.end_time) {
+            return
+        };
+        let total_stake = coin::value(&pool.stake);
+        if (total_stake == 0) {
+            pool.last_update_time = now;
+            return
+        };
+        let duration_us = if (now <= pool.end_time) {
+            now - pool.last_update_time
+        } else {
+            pool.end_time - pool.last_update_time
+        };
+
+        // total reward since last pool update
+        let duration_reward = (duration_us as u128) * (pool.reward_remaining as u128) / (pool.end_time - pool.last_update_time as u128);
+        // add reward_per_share since last update to acc_reward_per_share
+        pool.acc_reward_per_share = pool.acc_reward_per_share + duration_reward * REWARD_PER_SHARE_MUL / (total_stake as u128);
```

Comment:
since this rounding down, is it correct that total reward will be smaller or equal than `duration_reward`?